src/main/java/com/google/jenkins/plugins/computeengine/client/ComputeClient.java

```diff
@@ -418,6 +425,85 @@ public void deleteTemplate(String projectId, String templateName) throws IOExcep
+     */
+    public void createSnapshot(String projectId, String zone, String instanceId) throws IOException, InterruptedException {
+        try {
+            zone = nameFromSelfLink(zone);
+            Instance instance = compute.instances().get(projectId, zone, instanceId).execute();
+
+            //TODO: JENKINS-56113 parallelize snapshot creation
+            for (AttachedDisk disk : instance.getDisks()) {
+                    String diskId = nameFromSelfLink(disk.getSource());
+                    createSnapshotForDisk(projectId, zone, diskId);
+            }
+        } catch (InterruptedException ie) {
+            // catching InterruptedException here because calling function also can throw InterruptedException from trying to terminate node
+            LOGGER.log(Level.WARNING,"Error in creating snapshot.", ie);
+            throw ie;
+        } catch (IOException ioe) {
+            LOGGER.log(Level.WARNING,"Interruption in creating snapshot.", ioe);
+            throw ioe;
+        }
+    }
+
+    /**
+     * Given a disk's name, create a snapshot for said disk.
+     *
+     * @param projectId Google cloud project id (e.g. my-project).
+     * @param zone Zone of disk.
+     * @param diskId Name of disk to create a snapshot for.
+     *
+     * @throws IOException If an error occured in snapshot creation.
+     * @throws InterruptedException If snapshot creation is interrupted.
+     */
+    public void createSnapshotForDisk(String projectId, String zone, String diskId) throws IOException, InterruptedException {
+        Snapshot snapshot = new Snapshot();
+        snapshot.setName(diskId);
+
+        Operation op = compute.disks().createSnapshot(projectId, zone, diskId, snapshot).execute();
+        // poll for result
+        waitForOperationCompletion(projectId, op.getName(), op.getZone(), SNAPSHOT_TIMEOUT_MILLISECONDS);
+    }
+
+    /**
+     * Deletes persistent disk snapshot. Does not block.
+     *
+     * @param projectId Google cloud project id (e.g. my-project).
+     * @param snapshotName Name of the snapshot to be deleted.
+     * @throws IOException If an error occurred in deleting the snapshot.
+     */
+    public void deleteSnapshot(String projectId, String snapshotName) throws IOException {
+        compute.snapshots().delete(projectId, snapshotName).execute();
+    }
+
+    /**
+     * Returns snapshot with name snapshotName
+     *
+     * @param projectId Google cloud project id (e.g. my-project).
+     * @param snapshotName Name of the snapshot to get.
+     * @return Snapshot object with given snapshotName. Null if not found.
+     * @throws IOException If an error occurred in retrieving the snapshot.
+     */
+    public Snapshot getSnapshot(String projectId, String snapshotName) throws IOException {
+        try {
+            return compute.snapshots().get(projectId, snapshotName).execute();
+        } catch (GoogleJsonResponseException e) {
```

Comment:
Suggest not depending on an undocumented exception being thrown to handle the missing case. Instead consider using the list method for reliability, see: https://developers.google.com/resources/api-libraries/documentation/compute/v1/java/latest/com/google/api/services/compute/Compute.Snapshots.html#list-java.lang.String-

Reply:
I just don’t think it makes sense to list everything since we could have to go through a lot of snapshots and that’s probably the underlying code for snapshot.get. The list method involves going through several pages with a pagination token, making more room for error.

Reply:
The risk you run when depending on an undocumented behavior like this, is that it could change without notice which would break this function.

Reply:
Though I don't completely see eye to eye on this, I agree it will be nice to have to not fix the test if the API does change. Therefore, done.